File: apps/web/src/types/recipe.ts
/**
 * TypeScript types for the FitTrack AI recipes domain.
 *
 * Mirrors the Pydantic schemas in apps/api/app/schemas/recipe.py
 */

import type { FoodLogEntry, MacroTotals, MealType } from "@/types/nutrition";

export interface RecipeItem {
  food_id: string;
  food_name: string;
  food_brand: string | null;
  quantity_g: number;
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number | null;
}

export interface Recipe {
  id: string;
  name: string;
  description: string | null;
  items: RecipeItem[];
  totals: MacroTotals;
  created_at: string;
  updated_at: string;
}

export interface RecipeListResponse {
  recipes: Recipe[];
  total: number;
  page: number;
  page_size: number;
}

export interface RecipeItemInput {
  food_id: string;
  quantity_g: number;
}

export interface CreateRecipePayload {
  name: string;
  description?: string;
  items: RecipeItemInput[];
}

export interface UpdateRecipePayload {
  name?: string;
  description?: string;
  items?: RecipeItemInput[];
}

export interface LogRecipePayload {
  logged_date: string; // YYYY-MM-DD
  meal_type?: MealType;
  scale_factor?: number; // 1.0 = log exactly as saved
}

export interface LogRecipeResult {
  entries: FoodLogEntry[];
  totals: MacroTotals;
}

File: apps/api/app/schemas/recipe.py
"""Schemas for the recipes feature.

A recipe is a named, user-owned combination of foods + exact quantities
that can be re-logged later - optionally scaled - without re-describing it
or re-running AI estimation. Macro figures on items/totals are always
computed deterministically from the underlying Food's per-100g values,
never stored redundantly on the recipe itself (a Food's macros can be
edited later; the recipe should reflect current values, same as FoodLog).
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.schemas.nutrition import FoodLogResponse, MacroTotals, MealType

# ── Recipe items ──────────────────────────────────────────────────────────────


class RecipeItemInput(BaseModel):
    food_id: uuid.UUID
    quantity_g: float = Field(gt=0)


class RecipeItemResponse(BaseModel):
    food_id: uuid.UUID
    food_name: str
    food_brand: str | None
    quantity_g: float

    # Computed macros for this item at its saved quantity.
    calories: float
    protein_g: float
    carbs_g: float
    fat_g: float
    fiber_g: float | None


# ── Recipe CRUD ───────────────────────────────────────────────────────────────


class CreateRecipeRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=1000)
    items: list[RecipeItemInput] = Field(min_length=1, max_length=50)


class UpdateRecipeRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    # When provided, replaces the entire item list.
    items: list[RecipeItemInput] | None = Field(default=None, min_length=1, max_length=50)


class RecipeResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    items: list[RecipeItemResponse]
    totals: MacroTotals
    created_at: datetime
    updated_at: datetime


class RecipeListResponse(BaseModel):
    recipes: list[RecipeResponse]
    total: int
    page: int
    page_size: int


# ── Logging a recipe ──────────────────────────────────────────────────────────


class LogRecipeRequest(BaseModel):
    logged_date: date
    meal_type: MealType = "other"
    # 1.0 = log exactly as saved. E.g. 0.5 logs half the saved quantities.
    scale_factor: float = Field(default=1.0, gt=0, le=20)


class LogRecipeResponse(BaseModel):
    entries: list[FoodLogResponse]
    totals: MacroTotals
